process/data_processor.py

```diff
@@ -0,0 +1,155 @@
+"""
+loads and preprocesses the structured log data for anomoly prediction
+"""
+import numpy as np
+import pandas as pd
+import time
+import re
+from collections import OrderedDict
+from collections import Counter
+from scipy.special import expit
+
+
+def collect_event_ids(data_frame):
+    """
+    turns input data_frame into a 2 columned dataframe
+    with columns: BlockId, EventSequence
+    where EventSequence is a list of the events that happened to the block
+    """
+    data_dict = OrderedDict()
+    for _, row in data_frame.iterrows():
+        blk_id_list = re.findall(r"(blk_-?\d+)", row["Content"])
+        blk_id_set = set(blk_id_list)
+        for blk_id in blk_id_set:
+            if not blk_id in data_dict:
+                data_dict[blk_id] = []
+            data_dict[blk_id].append(row["EventId"])
+    data_df = pd.DataFrame(
+        list(data_dict.items()), columns=["BlockId", "EventSequence"]
+    )
+    return data_df
+
+
+class FeatureExtractor(object):
+    """
+    class for fitting and transforming the training set
+    then transforming the testing set
+    """
+
+    def __init__(self):
+        self.mean_vec = None
+        self.idf_vec = None
+        self.events = None
+        self.term_weighting = None
+        self.normalization = None
+
+    def fit_transform(self, X_seq, term_weighting=None, normalization=None):
+        """
+        Fit and transform the training set
+        X_Seq: ndarray,  log sequences matrix
+        term_weighting: None or `tf-idf`
+        normalization: None - not implemented yet
+        """
+        self.term_weighting = term_weighting
+        self.normalization = normalization
+
+        # Convert into bag of words
+        X_counts = []
+        for i in range(X_seq.shape[0]):
+            event_counts = Counter(X_seq[i])
+            X_counts.append(event_counts)
+        X_df = pd.DataFrame(X_counts)
+        X_df = X_df.fillna(0)
+        self.events = X_df.columns
+        X = X_df.values
+
+        num_instance, num_event = X.shape
+        # applies tf-idf if pararmeter
+        if self.term_weighting == "tf-idf":
+            df_vec = np.sum(X > 0, axis=0)
+            self.idf_vec = np.log(num_instance / (df_vec + 1e-8))
+            idf_matrix = X * np.tile(self.idf_vec, (num_instance, 1))
+            X = idf_matrix
+
+        # normalization if parameter
+        if self.normalization == "zero-mean":
```

Comment:
Suggest this just gets removed as we're not currently using it (or planning on using it). It also appears that it's just centering the data at 0, not actually normalizing or standardizing. We can always do normalization as a preprocessing step of the data frame as required as well. (although we'll need to think about how that gets done if we have a matrix of input for a CNN).

Similarly, suggest removing the sigmoid normalization as I think we should only keep what we know / are using in the code.

Reply:
Sure , can remove